include/lsst/afw/math/polynomials.h

```diff
@@ -0,0 +1,96 @@
+
+/**
+ *  @namespace lsst::afw::math::polynomials Low-level polynomials (including special polynomials) in C++.
+ *
+ *  The afw::math::polynomials library provides low-level classes for
+ *  efficiently evaluating polynomial basis functions and expansions in C++.
+ *  The classes here:
+ *   - are not available in Python;
+ *   - are not polymorphic (no virtual functions);
+ *   - provide workspace objects to minimize memory allocations when
+ *     appropriate;
+ *   - do not throw exceptions (users are responsible for providing valid
+ *     inputs);
+ *   - have almost no outside dependencies (just Eigen);
+ *   - use templates to allow them to work with any array/vector objects (not
+ *     just Eigen).
+ *
+ *  They are intended to be used as the building blocks of higher-level
+ *  objects that are visible to Python users and persisted with our data
+ *  products, such as ChebyshevBoundedField and the afw::math::Function
+ *  hierarchy (DM-13156 and DM-13157, respectively).
+ *
+ *  At present, the library only includes support for 1-d and 2-d standard
+ *  polynomials and Chebyshev polynomials of the first kind, but adding
+ *  support for any other function defined by a recurrence relation (i.e. any
+ *  other special polynomial) should be extremely easy (see
+ *  RecurrenceBasis1d), and need not be done within the polynomials library
+ *  itself.
+ *
+ *  For both 1-d and 2-d, the library contains the following kinds of objects:
+ *
+ *   - Basis1d and Basis2d: objects that evaluate basis functions at one
+ *     point at a time.  The fundamental 1-d basis is the RecurrenceBasis1d
+ *     template, while the fundamental 2-d basis is the PackedBasis2d template.
+ *     @ref PolynomialBasis1d, @ref Chebyshev1Basis1d, @ref PolynomialBasis2d, and
+ *     @ref Chebyshev1Basis2d provide typedefs to instantiations of these that
+ *     should generally be preferred.
+ *
+ *   - Function1d and Function2d: templates that combine a Basis1d or Basis2d
+ *     with a vector of coefficients.
+ *
+ *   - Scaling1d and Scaling2d: transformation objects that, with the
+ *     ScaledBasis1d and ScaledBasis2d templates, allow a basis or function to
+ *     be constructed that remaps points as part of evaluating the basis
+ *     functions.  For example, a ScaledChebyshev1Basis1d combines both the a
+ *     Chebyshev polynomial basis and the scaling from some domain to [-1, 1]
+ *     that enables most special properties of Chebyshevs.  When necessary,
+ *     the @ref simplify functions can be used to convert a scaled polynomial
+ *     function to an equivalent unscaled function.
+ *
+ *  The library also includes a few utility classes and functions:
+ *
+ *   - BinomialMatrix, which provides an efficient and stable way to get
+ *     binomial coefficients.
+ *
+ *   - PackedIndexIterator and PackedIndexRange, which handle the flattening
+ *     of pairs of 1-d polynomials coefficients and basis functions into 1-d
+ *     arrays.
+ */
```

Comment:
I strongly suspect you should put this code (including useful functions based on it) in a new package, rather than `afw`, since it seems potentially generally useful and has a minimum of dependencies. I'm thinking along the lines of your RFC for using sphgeom and I'd like to avoid a change like that in the future.

Reply:
Working on this was part of the inspiration for that RFC, but because some of this code *does* depend on `afw.geom` it can't be moved out of afw yet, and I'd rather not split it up based on that dependency problem; I'd rather address the dependency problem by moving both the geometry primitives and this code to another package.  But since that's a lot more work, I can't make a good case that it's what LSST should spend its time on, and hence the RFC.